### ISSUE TITLE
proposal: use npm-shrinkwrap instead of pinning exact versions

### ISSUE THREAD

**Issue by aliceriot** [mitodl/micromasters#399]
TL;DR: We've seen issues even with pinning exact versions of npm packages. I think using [npm-shrinkwrap](https://docs.npmjs.com/cli/shrinkwrap) could help us avoid this in the future.

## Details

An `npm-shrinkwrap.json` file works much like a `Gemfile.lock` file does in Ruby/bundler. Every time a dependency is added `npm` will automatically update the shrinkwrap file to reflect the addition of the dependency, and when you do `npm install` it will decide what to install (and at which versions) based on the shrinkwrap file, instead of looking at the allowed versions in the `package.json` and deciding what to do from there (something `npm` is notoriously, shall we say, creative with).

I've used it on some side projects - here's an example: https://github.com/aliceriot/literate-crypto/blob/master/npm-shrinkwrap.json - and found it really helpful.

The shrinkwrap file pins exact versions for our explicit dependencies (which we currently pin in `package.json`) and also for any dependencies that those dependencies pull in. The problem we saw today (5/23/16) was due to a new version of React being pushed, which our dependencies allowed (because they specify a range). Using npm shrinkwrap would circumvent this problem, because we pin exact dependency-of-dependency versions, so newer version of React (which meet the dependency version requirements of our dependencies) won't cause the versions of packages being pulled in to change.

This should give us more confidence in `npm install` doing the same (or at least, a really similar) thing every time.

## Implementation of proposal

If we decide to do this I think we'll need to bump our node / npm version, shrinkwrap support didn't get that great until npm 3.

Then there's only one other catch: `npm shrinkwrap` doesn't include `devDependencies` by default. I'd suggest two options here:

1. we can pass the `--dev` flag to `npm shrinkwrap` to include dev dependencies in the shrinkwrap file
2. we can discard `devDependencies` altogether and put all dependencies in `dependencies`.

Personally I favor 2, it's more simple overall, and (as far as I can tell) we aren't installing that much extra stuff in `devDependencies` anyhow.

Then switching to `npm-shrinkwrap` would look basically like:

1. `rm -rf node_modules`
2. reorganize `devDependencies`
3. bump node version so we can use npm 3
4. `npm install`
5. `npm shrinkwrap`
6. commit `npm-shrinkwrap.json`

Anyway, feedback?

**Comment by gsidebo** [comment]
👍

**Comment by noisecapella** [comment]
Overall using npm shrinkwrap is probably a good idea. I'm wary about mixing devDependencies into dependencies because I'm not sure how it impacts the bundle file size. How would this work when we install a new dependency? Run `npm install` then `npm shrinkwrap`?

**Comment by aliceriot** [comment]
With npm 3 the shrinkwrap will be updated anytime you pass `--save`, so if you need to add e.g. `material-ui` you would just do `npm install --save material-ui`, and `npm` would insert `material-ui` into `package.json` and add a statement for `material-ui` to the shrinkwrap file.

The only thing that I'm a little unsure of is how ongoing maintenance of the shrinkwrap will work if dev dependencies stay separate. The bundle size concern is real though - once webpack 2 (with tree shaking and all that) lands it probably won't be a concern.

Do you think this is worth some investigation? We could just experiment, like move all of our current `devDependencies` into `dependencies` and see how big the output of `./webpack_if_prod.sh` is. I'm willing to do a little fiddling to add some data to the discussion!

**Comment by aliceriot** [comment]
Some testing!

First off, I wanted to get a baseline estimate for how big our current JS bundle is. Here's the output of running `node node_modules/webpack/bin/webpack.js --config webpack.config.prod.js --bail --progress` in the `watch` container:

![before_node_update](https://cloud.githubusercontent.com/assets/6207644/15505983/ba711838-2193-11e6-9ab8-8769e45f9716.png)

after updating the nodejs version to 6.2 (and npm to 3.8.9):

![node_6 2](https://cloud.githubusercontent.com/assets/6207644/15506559/24885f68-2196-11e6-9583-8871e1a9e947.png)

so that's exactly the same. Then I moved all the `devDependencies` into `dependencies`, and ran the same command:

![dev_dependencies_integrated](https://cloud.githubusercontent.com/assets/6207644/15507111/35891f6c-2198-11e6-8732-384cafb694d2.png)

so it seems that even just sticking everything in `dependencies` we end up with the same bundle sizes. This makes sense - we aren't explicitly including the development dependencies (webpack, babel, etc) when we build production bundles anyway, so they shouldn't increase the size.

**Comment by aliceriot** [comment]
curious to hear your thoughts @noisecapella. I have a branch with the changes that I can push up to github as well if you'd like to poke at it.

**Comment by noisecapella** [comment]
Sure, that would be great. I'm fine with moving everything to dependencies, it would also prevent problems where we install something as a devDependency and don't realize it's needed in production until it gets onto one of the deploy servers

**Comment by aliceriot** [comment]
One other question: currently we're installing nodejs and npm inside of the web container (which we build with a dockerfile). Is there a reason we're doing this instead of using one of the prebuilt nodejs containers for `watch`? e.g. https://hub.docker.com/_/node/

Instead, we could, in `docker-compose.yml`, declare that `watch` uses the `node:6.2` image - then we'd avoid having to install node and npm in the web container. We'd need to tweak `travis-docker-compose.yml` to also have a `watch` container, and add a command to `.travis.yml` to run the JS tests in the `watch` container.

I think it would take a little fiddling to move over to that, but it might reduce the startup time for our tests (since preparing the JS environment would mean just using a prebuilt image, which travis probably already has cached).

**Comment by noisecapella** [comment]
There's no reason not to do things that way. In LORE we put node and npm inside the web container because django-compressor ran the JSX transformer inside the same container. We should be able to get rid of node and npm and do this in a separate container.

**Comment by aliceriot** [comment]
Cool, maybe I'll take a poke at doing that in the branch I have open, and see if it's easy to get it to work. I'd looooove to get the CI test time down, it's a lot of waiting for packages to install right now.

**Comment by noisecapella** [comment]
If we do this we will also need to change our documentation for running unit tests. What would we replace `docker-compose run web tox` with?

**Comment by aliceriot** [comment]
`docker-compose run watch ./js_test.js`? we could add a little file that just has:

```js
#!/usr/bin/env node

npm run-script coverage
npm run-script lint
```

alternatively, if we wanted one command to run them all we could just have a small shellscript:

```bash
#!/usr/bin/env bash

docker-compose run web tox
docker-compose run watch npm run-script coverage
docker-compose run watch npm run-script lint
```

**Comment by noisecapella** [comment]
That works for me as long as the documentation gets updated and whatever changes we make are communicated to the group

**Comment by aliceriot** [comment]
cool!

**Comment by gsidebo** [comment]
i'd prefer it if said shell script would run js tests only with tox, or at least have the option to do so

**Comment by noisecapella** [comment]
Why do you prefer it that way? Are you just looking for a single place to run all the tests?

**Comment by gsidebo** [comment]
just realized i was assuming that coverage and lint were specific to our JS code. if that's not true, ignore me 😄

**Comment by noisecapella** [comment]
FYI There is linting and coverage for both python and JS. In Python it's done with pylint and coverage using pytest plugins (see pytest.ini). In JS we are running them as separate commands `npm run lint` and `npm run coverage`. The coverage output for Python is at `htmlcov/` and for JS it's at `coverage/`

**Comment by aliceriot** [comment]
FYI this is what I have currently (over in #402), `test_suite.sh`:

```bash
#!/usr/bin/env bash

docker-compose run web tox
docker-compose run watch npm run-script lint
docker-compose run watch npm run-script coverage
```

so you can do `./test_suite.sh` to run all the tests.

**Comment by aliceriot** [comment]
@gsidebo I'm not exactly sure what you mean, are you asking whether `tox` will still run the JS tests?

**Comment by aliceriot** [comment]
...oops
